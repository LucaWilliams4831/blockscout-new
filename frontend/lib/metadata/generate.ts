import type { Route } from 'nextjs-routes';

import type { ApiData, Metadata } from './types';

import config from 'configs/app';
import getNetworkTitle from 'lib/networks/getNetworkTitle';

import compileValue from './compileValue';
import * as templates from './templates';

export default function generate<R extends Route>(route: R, apiData?: ApiData<R>): Metadata {
  const params = {
    ...route.query,
    ...apiData,
    network_name: config.chain.name,
    network_title: getNetworkTitle(),
  };

  const title = compileValue(templates.title.make(route.pathname), params);
  const description = compileValue(templates.description.make(route.pathname), params);

  return {
    title,
    description,
  };
}
