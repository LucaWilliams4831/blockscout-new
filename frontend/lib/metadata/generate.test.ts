import type { Route } from 'nextjs-routes';

import type { ApiData } from './types';

import generate from './generate';

interface TestCase<R extends Route> {
  title: string;
  route: R;
  apiData?: ApiData<R>;
}

const TEST_CASES: Array<TestCase<Route>> = [
  {
    title: 'static route',
    route: {
      pathname: '/blocks',
    },
  },
  {
    title: 'dynamic route',
    route: {
      pathname: '/tx/[hash]',
      query: { hash: '0x12345' },
    },
  },
  {
    title: 'dynamic route with API data',
    route: {
      pathname: '/token/[hash]',
      query: { hash: '0x12345' },
    },
    apiData: { symbol: 'USDT' },
  } as TestCase<{ pathname: '/token/[hash]'; query: { hash: string }}>,
];

describe('generates correct metadata for:', () => {
  TEST_CASES.forEach((testCase) => {
    it(`${ testCase.title }`, () => {
      const result = generate(testCase.route, testCase.apiData);
      expect(result).toMatchSnapshot();
    });
  });
});
