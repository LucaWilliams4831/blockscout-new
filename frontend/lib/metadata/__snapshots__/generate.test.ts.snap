// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generates correct metadata for: dynamic route 1`] = `
{
  "description": "View transaction 0x12345 on Blockscout (Blockscout) Explorer",
  "title": "Blockscout transaction 0x12345 | Blockscout",
}
`;

exports[`generates correct metadata for: dynamic route with API data 1`] = `
{
  "description": "0x12345, balances and analytics on the Blockscout (Blockscout) Explorer",
  "title": "Blockscout USDT token details | Blockscout",
}
`;

exports[`generates correct metadata for: static route 1`] = `
{
  "description": "Blockscout is the #1 open-source blockchain explorer available today. 100+ chains and counting rely on Blockscout data availability, APIs, and ecosystem tools to support their networks.",
  "title": "Blockscout blocks | Blockscout",
}
`;
